tcms/rpc/tests/test_testplan.py

```diff
@@ -226,3 +228,123 @@ def test_ignores_existing_mappings(self):
+        self.assertEqual(self.params['product_version'], plan['product_version_id'])
+        self.assertEqual(self.params['name'], plan['name'])
+        self.assertEqual(self.params['type'], plan['type_id'])
+        self.assertEqual(self.params['text'], plan['text'])
+        self.assertEqual(self.params['parent'], plan['parent'])
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegex(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestPlan.create(self.params)
+
+
+class TestListAttachments(APITestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+    def test_list_attachments_with_wrong_plan_id(self):
+        with self.assertRaisesRegex(XmlRPCFault, 'TestPlan matching query does not exist'):
+            self.rpc_client.TestPlan.list_attachments(-1)
+
+
+class TestListAttachmentsPermissions(APIPermissionsTestCase):
+    permission_label = 'attachments.view_attachment'
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.plan = TestPlanFactory()
+
+    def verify_api_with_permission(self):
+        attachments = self.rpc_client.TestPlan.list_attachments(self.plan.pk)
+        self.assertEqual(0, len(attachments))
+
+    def verify_api_without_permission(self):
+        with self.assertRaisesRegexp(ProtocolError, '403 Forbidden'):
+            self.rpc_client.TestPlan.list_attachments(self.plan.pk)
+
+
+class TestAddAttachment(APITestCase):
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.plan = TestPlanFactory()
+
+    def test_add_attachment(self):
+        self.rpc_client.TestPlan.add_attachment(self.plan.pk, 'attachment.txt', 'kiwitcms')
+        attachments = self.rpc_client.TestPlan.list_attachments(self.plan.pk)
+        self.assertEqual(1, len(attachments))
+
+        attachment = attachments[0]
+        self.assertEqual(self.api_user.pk, attachment['owner_pk'])
+        self.assertEqual(self.api_user.username, attachment['owner_username'])
+
+        file_url = attachment['url']
+        fn = file_url[file_url.rindex('/') + 1:]
+        self.assertTrue(fn.startswith('attachment'))
+        self.assertTrue(fn.endswith('.txt'))
+
+
+class TestAddAttachmentPermissions(APIPermissionsTestCase):
+    permission_label = 'attachments.add_attachment'
+
+    def _fixture_setup(self):
+        super()._fixture_setup()
+
+        self.plan = TestPlanFactory()
+
+    def verify_api_with_permission(self):
+        self.rpc_client.TestPlan.add_attachment(self.plan.pk, 'attachment.txt', 'kiwitcms')
+        attachments = Attachment.objects.attachments_for_object(self.plan)
+        self.assertEqual(1, len(attachments))
+
+        attachment = attachments[0]
+        file_url = attachment.attachment_file.url
+        fn = file_url[file_url.rindex('/') + 1:]
+        self.assertTrue(fn.startswith('attachment'))
+        self.assertTrue(fn.endswith('.txt'))
```

Comment:
Very similar comments to above. In fact in this particular case you can merge the functional and permissions scenarios because they are the same and there are no if conditions/various scenarios for the functionality part.